plutus-core/executables/Common.hs

```diff
@@ -62,35 +66,170 @@ import           Text.Read                                (readMaybe)
    the simplest thing to do and fits our use case.
  -}
 
+----------- Executable type class -----------
+-- currently we only have PLC and UPLC. PIR will be added later on.
+
+-- | PLC program type.
+type PlcProg =
+  PLC.Program PLC.TyName PLC.Name PLC.DefaultUni PLC.DefaultFun
+
+-- | UPLC program type.
+type UplcProg =
+  UPLC.Program PLC.Name PLC.DefaultUni PLC.DefaultFun
+
+class Executable a where
+
+  -- | Parse a program.
+  parseProgram ::
+    (AsParseError e PLC.AlexPosn, MonadError e m, PLC.MonadQuote m) =>
+    BSL.ByteString ->
+      m (a PLC.AlexPosn)
+
+  -- | Check a program for unique names.
+  -- Throws a @UniqueError@ when not all names are unique.
+  checkProgram ::
+     (Ord ann, AsUniqueError e ann,
+       MonadError e m)
+    => (UniqueError ann -> Bool)
+    -> a ann
+    -> m ()
+
+  deserialiseRestoringUnitsOrFail ::
+    BSL.ByteString
+      -> Either DeserialiseFailure (a ())
```

Comment:
`PlcProg` sometimes take `AlexPosn` and sometimes `()`. That's why there's the `(a b)` annotations because I was thinking in terms of e.g. `PlcProg AlexPosn`.